core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1080,6 +1080,21 @@ abstract class RDD[T: ClassTag](
       var partiallyAggregated = mapPartitions(it => Iterator(aggregatePartition(it)))
       var numPartitions = partiallyAggregated.partitions.length
       val scale = math.max(math.ceil(math.pow(numPartitions, 1.0 / depth)).toInt, 2)
+
+      // Do one level of aggregation based on executorId before starting the tree
+      // NOTE: exclude the driver from list of executors
+      val numExecutors = math.max(context.getExecutorStorageStatus.length - 1, 1)
+      partiallyAggregated = partiallyAggregated.mapPartitionsWithIndex { case (idx, iter) =>
+        def isAllDigits(x: String) = x forall Character.isDigit
+        val execId = SparkEnv.get.executorId
+        if (isAllDigits(execId)) {
+          iter.map((execId.toInt, _))
+        } else {
+          iter.map((execId.hashCode, _))
```

Comment:
I think executor ID is always an integer except maybe during tests, but given that it's a string everywhere that assumption may not be correct, and we should certainly not fail when that happens.

Actually why do we need to convert it to an int? As long as they are distinct then reduceByKey should do the right thing, no?